sql/opt/norm: TestNormRules failed under stress

SHA: https://github.com/cockroachdb/cockroach/commits/59792cfdedad10828fcbf6328fe4195b6de5d8f0

Parameters:
```
TAGS=
GOFLAGS=-race
```

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=TestNormRules PKG=github.com/cockroachdb/cockroach/pkg/sql/opt/norm TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1051538&tab=buildLog

```

 │    │    ├── columns: x:6(int!null) y:7(int)
 │    │    ├── key: (6)
 │    │    └── fd: (6)-->(7)
 │    └── filters (true)
 └── filters
      └── ((k = x) OR (k IS NULL)) OR (x IS NULL) [type=bool, outer=(1,6)]

testdata/rules/select:812:
SELECT * FROM a INNER JOIN xy ON a.k=xy.x WHERE a.f=1.1 AND s='foo' AND xy.y=10 AND a.i<xy.y
----
inner-join (lookup a)
 ├── columns: k:1(int!null) i:2(int!null) f:3(float!null) s:4(string!null) j:5(jsonb) x:6(int!null) y:7(int!null)
 ├── key columns: [6] = [1]
 ├── key: (6)
 ├── fd: ()-->(3,4,7), (1)-->(2,5), (1)==(6), (6)==(1)
 ├── select
 │    ├── columns: x:6(int!null) y:7(int!null)
 │    ├── key: (6)
 │    ├── fd: ()-->(7)
 │    ├── scan xy
 │    │    ├── columns: x:6(int!null) y:7(int)
 │    │    ├── key: (6)
 │    │    └── fd: (6)-->(7)
 │    └── filters
 │         └── y = 10 [type=bool, outer=(7), constraints=(/7: [/10 - /10]; tight), fd=()-->(7)]
 └── filters
      ├── i < y [type=bool, outer=(2,7), constraints=(/2: (/NULL - ]; /7: (/NULL - ])]
      ├── f = 1.1 [type=bool, outer=(3), constraints=(/3: [/1.1 - /1.1]; tight), fd=()-->(3)]
      └── s = 'foo' [type=bool, outer=(4), constraints=(/4: [/'foo' - /'foo']; tight), fd=()-->(4)]

testdata/rules/select:835:
SELECT * FROM a, xy WHERE a.i=100 AND $1>'2000-01-01T1:00:00' AND xy.x=a.k
----
inner-join (lookup xy)
 ├── columns: k:1(int!null) i:2(int!null) f:3(float) s:4(string) j:5(jsonb) x:6(int!null) y:7(int)
 ├── key columns: [1] = [6]
 ├── has-placeholder
 ├── key: (6)
 ├── fd: ()-->(2), (1)-->(3-5), (6)-->(7), (1)==(6), (6)==(1)
 ├── select
 │    ├── columns: k:1(int!null) i:2(int!null) f:3(float) s:4(string) j:5(jsonb)
 │    ├── has-placeholder
 │    ├── key: (1)
 │    ├── fd: ()-->(2), (1)-->(3-5)
 │    ├── scan a
 │    │    ├── columns: k:1(int!null) i:2(int) f:3(float) s:4(string) j:5(jsonb)
 │    │    ├── key: (1)
 │    │    └── fd: (1)-->(2-5)
 │    └── filters
 │         ├── $1 > '2000-01-01T1:00:00' [type=bool]
 │         └── i = 100 [type=bool, outer=(2), constraints=(/2: [/100 - /100]; tight), fd=()-->(2)]
 └── filters
      └── $1 > '2000-01-01T1:00:00' [type=bool]

testdata/rules/select:864:
SELECT * FROM (SELECT i, count(*) FROM a GROUP BY i) a WHERE i=1
----
group-by
 ├── columns: i:2(int) count:6(int)
 ├── cardinality: [0 - 1]
 ├── key: ()
 ├── fd: ()-->(2,6)
 ├── select
 │    ├── columns: i:2(int!null)
 │    ├── fd: ()-->(2)
 │    ├── scan a
 │    │    └── columns: i:2(int)
 │    └── filters
 │         └── i = 1 [type=bool, outer=(2), constraints=(/2: [/1 - /1]; tight), fd=()-->(2)]
 └── aggregations
      ├── count-rows [type=int]
      └── const-agg [type=int, outer=(2)]
           └── variable: i [type=int]

testdata/rules/select:885:
SELECT * FROM (SELECT i FROM a GROUP BY i) a WHERE i=1
----
distinct-on
 ├── columns: i:2(int)
 ├── cardinality: [0 - 1]
 ├── key: ()
 ├── fd: ()-->(2)
 ├── select
 │    ├── columns: i:2(int!null)
 │    ├── fd: ()-->(2)
 │    ├── scan a
 │    │    └── columns: i:2(int)
 │    └── filters
 │         └── i = 1 [type=bool, outer=(2), constraints=(/2: [/1 - /1]; tight), fd=()-->(2)]
 └── aggregations
      └── const-agg [type=int, outer=(2)]
           └── variable: i [type=int]

panic during testdata/rules/select:905:
SELECT * FROM (SELECT k, i, max(s) m FROM a GROUP BY k, i) a WHERE i=k AND m='foo'
panic: ordering column group (1,2) contains non-equivalent columns (op scan) [recovered]
	panic: ordering column group (1,2) contains non-equivalent columns (op scan) [recovered]
	panic: ordering column group (1,2) contains non-equivalent columns (op scan)

goroutine 42 [running]:
testing.tRunner.func1(0xc4206dd950)
```

--- a/pkg/sql/opt/exec/execbuilder/testdata/orderby
+++ b/pkg/sql/opt/exec/execbuilder/testdata/orderby
@@ -765,11 +765,13 @@ CREATE TABLE xyz (x INT, y INT, z INT, INDEX(z,y))
 query TTTTT
 EXPLAIN (VERBOSE) SELECT * FROM xyz WHERE z=1 AND x=y ORDER BY x;
 ----
-filter           ·       ·                (x, y, z)              +y
- │               filter  x = y            ·                      ·
- └── index-join  ·       ·                (x, y, z)              +y
-      ├── scan   ·       ·                (y, z, rowid[hidden])  +y
-      │          table   xyz@xyz_z_y_idx  ·                      ·
-      │          spans   /1/!NULL-/2      ·                      ·
-      └── scan   ·       ·                (x, y, z)              ·
-·                table   xyz@primary      ·                      ·
+sort                  ·       ·                (x, y, z)              +x
+ │                    order   +x               ·                      ·
+ └── filter           ·       ·                (x, y, z)              ·
+      │               filter  x = y            ·                      ·
+      └── index-join  ·       ·                (x, y, z)              ·
+           ├── scan   ·       ·                (y, z, rowid[hidden])  ·
+           │          table   xyz@xyz_z_y_idx  ·                      ·
+           │          spans   /1/!NULL-/2      ·                      ·
+           └── scan   ·       ·                (x, y, z)              ·
+·                     table   xyz@primary      ·                      ·

--- a/pkg/sql/opt/norm/testdata/rules/select
+++ b/pkg/sql/opt/norm/testdata/rules/select
@@ -924,7 +924,7 @@ select
  │    │    │    ├── columns: k:1(int!null) i:2(int) s:4(string)
  │    │    │    ├── key: (1)
  │    │    │    ├── fd: (1)-->(2,4)
- │    │    │    └── ordering: +(1|2) [provided: +1]
+ │    │    │    └── ordering: +1
  │    │    └── filters
  │    │         └── i = k [type=bool, outer=(1,2), constraints=(/1: (/NULL - ]; /2: (/NULL - ]), fd=(1)==(2), (2)==(1)]
  │    └── aggregations

--- a/pkg/sql/opt/ordering/select.go
+++ b/pkg/sql/opt/ordering/select.go
@@ -17,6 +17,7 @@ package ordering
 import (
 	"github.com/cockroachdb/cockroach/pkg/sql/opt"
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/memo"
+	"github.com/cockroachdb/cockroach/pkg/sql/opt/props"
 	"github.com/cockroachdb/cockroach/pkg/sql/opt/props/physical"
 )
 
@@ -31,7 +32,32 @@ func selectBuildChildReqOrdering(
 	if childIdx != 0 {
 		return physical.OrderingChoice{}
 	}
-	return *required
+	return trimColumnGroups(required, &parent.(*memo.SelectExpr).Input.Relational().FuncDeps)
+}
+
+// trimColumnGroups removes columns from ColumnOrderingChoice groups as
+// necessary, so that all columns in each group are equivalent according to
+// the given FDs. It is used when the parent expression can have column
+// equivalences that the input expression does not (for example a Select with an
+// equality condition); the columns in a group must be equivalent at the level
+// of the operator where the ordering is required.
+func trimColumnGroups(
+	required *physical.OrderingChoice, fds *props.FuncDepSet,
+) physical.OrderingChoice {
+	res := *required
+	copied := false
+	for i := range res.Columns {
+		c := &res.Columns[i]
+		eqGroup := fds.ComputeEquivGroup(c.AnyID())
+		if !c.Group.SubsetOf(eqGroup) {
+			if !copied {
+				res = res.Copy()
+				copied = true
+			}
+			res.Columns[i].Group = c.Group.Intersection(eqGroup)
+		}
+	}
+	return res
 }
 
 func selectBuildProvided(expr memo.RelExpr, required *physical.OrderingChoice) opt.Ordering {

--- a/pkg/sql/opt/xform/testdata/rules/groupby
+++ b/pkg/sql/opt/xform/testdata/rules/groupby
@@ -861,13 +861,13 @@ memo (optimized, ~8KB, required=[presentation: array_agg:5])
  ├── G3: (projections)
  ├── G4: (select G6 G7) (select G8 G7) (select G9 G7)
  │    ├── [ordering: +(2|3) opt(4)]
- │    │    ├── best: (select G8="[ordering: +(2|3) opt(4)]" G7)
+ │    │    ├── best: (select G8="[ordering: +2 opt(4)]" G7)
  │    │    └── cost: 1079.12
  │    ├── [ordering: +(2|3),+4]
- │    │    ├── best: (select G9="[ordering: +(2|3),+4]" G7)
- │    │    └── cost: 1079.12
+ │    │    ├── best: (sort G4)
+ │    │    └── cost: 1080.04
  │    ├── [ordering: +(2|3)]
- │    │    ├── best: (select G8="[ordering: +(2|3)]" G7)
+ │    │    ├── best: (select G8="[ordering: +2]" G7)
  │    │    └── cost: 1079.12
  │    ├── [ordering: +4,+(2|3)]
  │    │    ├── best: (sort G4)
@@ -877,49 +877,49 @@ memo (optimized, ~8KB, required=[presentation: array_agg:5])
  │         └── cost: 1079.12
  ├── G5: (aggregations G10)
  ├── G6: (scan kuvw) (scan kuvw@uvw) (scan kuvw@wvu) (scan kuvw@vw) (scan kuvw@w)
- │    ├── [ordering: +(2|3) opt(4)]
+ │    ├── [ordering: +2 opt(4)]
  │    │    ├── best: (scan kuvw@uvw)
  │    │    └── cost: 1080.01
- │    ├── [ordering: +(2|3),+4]
- │    │    ├── best: (scan kuvw@vw)
- │    │    └── cost: 1080.01
- │    ├── [ordering: +(2|3)]
+ │    ├── [ordering: +2,+4]
+ │    │    ├── best: (sort G6)
+ │    │    └── cost: 1310.30
+ │    ├── [ordering: +2]
  │    │    ├── best: (scan kuvw@uvw)
  │    │    └── cost: 1080.01
- │    ├── [ordering: +4,+(2|3)]
- │    │    ├── best: (scan kuvw@wvu)
- │    │    └── cost: 1080.01
+ │    ├── [ordering: +4,+2]
+ │    │    ├── best: (sort G6)
+ │    │    └── cost: 1310.30
  │    └── []
  │         ├── best: (scan kuvw)
  │         └── cost: 1080.01
  ├── G7: (filters G11)
  ├── G8: (scan kuvw@uvw,constrained)
- │    ├── [ordering: +(2|3) opt(4)]
+ │    ├── [ordering: +2 opt(4)]
  │    │    ├── best: (scan kuvw@uvw,constrained)
  │    │    └── cost: 1069.21
- │    ├── [ordering: +(2|3),+4]
+ │    ├── [ordering: +2,+4]
  │    │    ├── best: (sort G8)
  │    │    └── cost: 1296.90
- │    ├── [ordering: +(2|3)]
+ │    ├── [ordering: +2]
  │    │    ├── best: (scan kuvw@uvw,constrained)
  │    │    └── cost: 1069.21
- │    ├── [ordering: +4,+(2|3)]
+ │    ├── [ordering: +4,+2]
  │    │    ├── best: (sort G8)
  │    │    └── cost: 1296.90
  │    └── []
  │         ├── best: (scan kuvw@uvw,constrained)
  │         └── cost: 1069.21
  ├── G9: (scan kuvw@vw,constrained)
- │    ├── [ordering: +(2|3) opt(4)]
- │    │    ├── best: (scan kuvw@vw,constrained)
- │    │    └── cost: 1069.21
- │    ├── [ordering: +(2|3),+4]
- │    │    ├── best: (scan kuvw@vw,constrained)
- │    │    └── cost: 1069.21
- │    ├── [ordering: +(2|3)]
- │    │    ├── best: (scan kuvw@vw,constrained)
- │    │    └── cost: 1069.21
- │    ├── [ordering: +4,+(2|3)]
+ │    ├── [ordering: +2 opt(4)]
+ │    │    ├── best: (sort G9)
+ │    │    └── cost: 1286.06
+ │    ├── [ordering: +2,+4]
+ │    │    ├── best: (sort G9)
+ │    │    └── cost: 1296.90
+ │    ├── [ordering: +2]
+ │    │    ├── best: (sort G9)
+ │    │    └── cost: 1286.06
+ │    ├── [ordering: +4,+2]
  │    │    ├── best: (sort G9)
  │    │    └── cost: 1296.90
  │    └── []
@@ -946,7 +946,7 @@ memo (optimized, ~8KB, required=[presentation: sum:5])
  ├── G3: (projections)
  ├── G4: (select G6 G7) (select G8 G7) (select G9 G7)
  │    ├── [ordering: +(2|3)]
- │    │    ├── best: (select G8="[ordering: +(2|3)]" G7)
+ │    │    ├── best: (select G8="[ordering: +2]" G7)
  │    │    └── cost: 1079.12
  │    ├── [ordering: +4]
  │    │    ├── best: (sort G4)
@@ -956,7 +956,7 @@ memo (optimized, ~8KB, required=[presentation: sum:5])
  │         └── cost: 1079.12
  ├── G5: (aggregations G10)
  ├── G6: (scan kuvw) (scan kuvw@uvw) (scan kuvw@wvu) (scan kuvw@vw) (scan kuvw@w)
- │    ├── [ordering: +(2|3)]
+ │    ├── [ordering: +2]
  │    │    ├── best: (scan kuvw@uvw)
  │    │    └── cost: 1080.01
  │    ├── [ordering: +4]
@@ -967,7 +967,7 @@ memo (optimized, ~8KB, required=[presentation: sum:5])
  │         └── cost: 1080.01
  ├── G7: (filters G11)
  ├── G8: (scan kuvw@uvw,constrained)
- │    ├── [ordering: +(2|3)]
+ │    ├── [ordering: +2]
  │    │    ├── best: (scan kuvw@uvw,constrained)
  │    │    └── cost: 1069.21
  │    ├── [ordering: +4]
@@ -977,9 +977,9 @@ memo (optimized, ~8KB, required=[presentation: sum:5])
  │         ├── best: (scan kuvw@uvw,constrained)
  │         └── cost: 1069.21
  ├── G9: (scan kuvw@vw,constrained)
- │    ├── [ordering: +(2|3)]
- │    │    ├── best: (scan kuvw@vw,constrained)
- │    │    └── cost: 1069.21
+ │    ├── [ordering: +2]
+ │    │    ├── best: (sort G9)
+ │    │    └── cost: 1286.06
  │    ├── [ordering: +4]
  │    │    ├── best: (sort G9)
  │    │    └── cost: 1286.06

